frame/reward-distribution/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn reward_config)]
+    pub type RewardDistributionConfigStorage<T: Config> =
+        StorageValue<_, RewardDistributionConfig, ValueQuery>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Distribution configuration has been updated.
+        DistributionConfigurationChanged(RewardDistributionConfig),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Sum of all rations must be one whole (100%)
+        InvalidDistributionConfiguration,
+    }
+
+    #[pallet::genesis_config]
+    pub struct GenesisConfig {
+        pub reward_config: RewardDistributionConfig,
+    }
+
+    #[cfg(feature = "std")]
+    impl Default for GenesisConfig {
+        fn default() -> Self {
+            Self {
+                reward_config: Default::default(),
+            }
+        }
+    }
+
+    #[pallet::genesis_build]
+    impl<T: Config> GenesisBuild<T> for GenesisConfig {
+        fn build(&self) {
+            assert!(self.reward_config.is_consistent());
+            RewardDistributionConfigStorage::<T>::put(self.reward_config.clone())
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Sets the reward distribution configuration parameters which will be used from next block reward distribution.
+        ///
+        /// It is mandatory that all components of configuration sum up to one whole (**100%**),
+        /// otherwise an error `InvalidDistributionConfiguration` will be raised.
+        ///
+        /// - `reward_distro_params` - reward distribution params
+        ///
+        /// Emits `DistributionConfigurationChanged` with config embeded into event itself.
+        ///
+        #[pallet::weight(T::WeightInfo::set_configuration())]
+        pub fn set_configuration(
+            origin: OriginFor<T>,
+            reward_distro_params: RewardDistributionConfig,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+
+            ensure!(
+                reward_distro_params.is_consistent(),
+                Error::<T>::InvalidDistributionConfiguration
+            );
+            RewardDistributionConfigStorage::<T>::put(reward_distro_params.clone());
+
+            Self::deposit_event(Event::<T>::DistributionConfigurationChanged(
+                reward_distro_params,
+            ));
+
+            Ok(().into())
+        }
+    }
+
+    impl<Moment, T: Config> OnTimestampSet<Moment> for Pallet<T> {
+        fn on_timestamp_set(_moment: Moment) {
+            let inflation = T::Currency::issue(T::RewardAmount::get());
+            Self::on_nonzero_unbalanced(inflation);
+        }
+    }
+
+    impl<T: Config> OnUnbalanced<NegativeImbalanceOf<T>> for Pallet<T> {
```

Comment:
Perhaps `OnUnbalanced` usage should be dropped since I don't want to expose this to the outside - pallet is responsible both for issuing and distributing rewards.